07_leukemia_patient/simulation_patient_cell_lines/workflow.sh

```diff
@@ -0,0 +1,389 @@
+wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/"$human_gtf".gz
+
+pigz --decompress *gz
+
+STAR --runMode genomeGenerate \
+     --runThreadN $NTHREADS_genome \
+     --genomeDir . \
+     --genomeFastaFiles $human_fa \
+     --sjdbGTFfile $human_gtf \
+     --sjdbOverhang 179
+
+# running STARsolo
+
+cd ..
+mkdir -p ./starsolo
+
+gzip $r1
+gzip $r2
+
+STAR --runThreadN $NTHREADS \
+     --genomeDir ./genome \
+     --readFilesCommand zcat \
+     --outFileNamePrefix ./starsolo/ \
+     --readFilesIn $r2.gz $r1.gz  \
+     --soloType CB_UMI_Complex \
+     --soloAdapterSequence AATGNNNNNNNNNCCAC \
+     --soloCBposition 2_-9_2_-1 2_4_2_12 2_17_2_25 \
+     --soloUMIposition 3_10_3_17 \
+     --soloUMIlen 8 \
+     --soloCellReadStats Standard \
+     --soloCBwhitelist $CL1 $CL2 $CL3 \
+     --soloCBmatchWLtype 1MM \
+     --soloCellFilter None \
+     --outSAMattributes NH HI AS nM NM MD jM jI MC ch CB UB gx gn sS CR CY UR UY\
+     --outSAMtype BAM SortedByCoordinate \
+     --quantMode GeneCounts \
+     --sjdbGTFfile $COMBINED_GTF_GENOME \
+     --sjdbOverhang 179 \
+     --limitBAMsortRAM 20000 * 1024 * 1024 \
+     --outSAMunmapped Within
+
+samtools index $STARSOLO_BAM
+
+# get unmapped reads and reads that mapped to BCR and to ABL and append their name to the readname
+
+## Step 1: get unmapped reads, reads in the BCR ABL slice, reads with no gx tag from .bam file and output in new .fastq file. Append the CB and UB tag to their read id. Deduplicate CB and UB.  
+
+# getting reads in BCR and ABL region and adding 1M bp to each side
+
+bcr_start=23179704
+bcr_end=23318037
+abl_start=130713016
+abl_end=130887675
+
+echo "getting read names and deduplicating"
+
+# deduplication is done based on first sorting by UB (28), CB (27) and 1 (read id). If the combination of UB and CB is unique as well as the read name, then the read is appended. 
+# if first if loop check if read is already present in .fastq file or not, otherwise not added. 
+
+samtools view -H $STARSOLO_BAM -@ $NTHREADS > header.txt
+
+samtools view "$STARSOLO_BAM" "chr22:22179704-24318037" -@ "$NTHREADS" \
+| cat header.txt - \
+| samtools view -@ "$NTHREADS" \
+| grep -v "CB:Z:-" \
+| grep -v "UB:Z:-" \
+| sort -k28,28 -k27,27 -k1,1 \
+| parallel --pipe --block 10M -j "$NTHREADS" \
+'awk -v seen="" -v current="" -v current_ub_cb="" -v seen_ub_cb="" -v current_ub="" -v seen_ub="" \
+  "BEGIN { OFS=\"\t\" } \
+  { \
+    current = \$1; \
+    current_ub_cb = \$28 \";\" \$27; \
+    current_ub = \$28; \
+    cmd = \"grep -q \" current \" r2.fastq\"; \
+    exit_code = system(cmd); \
+    if (exit_code == 0) { next; } \
+    if (current_ub != seen_ub && current_ub_cb != seen_ub_cb && current != seen) { \
+      print \"@\" current \";\" \$27 \";\" \$28 \"\\n\" \$10 \"\\n+\\n\" \$11; \
+    } \
+    seen = current; \
+    seen_ub_cb = current_ub_cb; \
+    seen_ub = current_ub; \
+  }"' >> r2.fastq
+
+echo "BCR extracted"
```

Comment:
couldn't you reuse the abl/bcr awk trick by samtools viewing the two ranges, hence avoiding doing the two steps separately?

Reply:
I mean like in https://www.biostars.org/p/49306/#49326